### PR TITLE
Stripe Terminal SDK update 3.3.1

### PR DESCRIPTION
**Pull request by bozidarsevo**
## Description

- Updated Stripe Terminal SDK to [version](https://github.com/stripe/stripe-terminal-ios/releases) 3.3.1 
- Discussion: pfoUAQ-hK-p2
- Plan is to also use "Added network and wallet to SCPCardPresentDetails." change to add these parameters to tracking events

## Changes in SDK from current version (3.1.0)

- Versions 3.3.1, 3.2.1 just have fixes in them

### 3.2.0
* Update: retrievePaymentIntent and retrieveSetupIntent no longer require a connected reader.
* Update: Adds [SCPCharge](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPCharge.html).authorizationCode to the sdk's SCPCharge model when it is available.
    * Note for internet reader integrations, this feature requires [reader software version](https://stripe.com/docs/terminal/readers/bbpos-wisepos-e#reader-software-version) 2.18 or later to be installed on your internet reader.
* Update: Added network and wallet to SCPCardPresentDetails.
* Update: Added paymentMethodId to SCPPaymentIntent.
* Update: The amount of time a reader can be used offline before needing to be activated online has been reduced to 30 days.
* Update: SCPReaderNetworkStatus has a new case (SCPReaderStatusUnknown) that represents the network status of an internet reader that was discovered while the SDK is operating in offline mode.

### 3.3.0
* New: [rebootReader](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(im)rebootReader:) method to reboot Bluetooth readers.
* New: An optional [reader:didDisconnect:](https://stripe.dev/stripe-terminal-ios/docs/Protocols/SCPBluetoothReaderDelegate.html#/c:objc(pl)SCPBluetoothReaderDelegate(im)reader:didDisconnect:) method was added to [SCPBluetoothReaderDelegate](https://stripe.dev/stripe-terminal-ios/docs/Protocols/SCPBluetoothReaderDelegate.html#/c:objc(pl)SCPBluetoothReaderDelegate) which provides a new [SCPDisconnectReason](https://stripe.dev/stripe-terminal-ios/docs/Enums/SCPDisconnectReason.html) to communicate known reasons why a Bluetooth reader disconnected.
* New: Support refunding payments with the SCPPaymentIntent.stripeId.
    * Note for internet reader integrations, this feature requires [reader software version](https://stripe.com/docs/terminal/readers/bbpos-wisepos-e#reader-software-version) 2.19 or later to be installed on your internet reader.
* New: Added support for retrieving and updating reader settings on WisePOS E and Stripe S700 by calling retrieveReaderSettings and setReaderSettings on SCPTerminal.
    * Beta: Accessibility settings are provided at this time, allowing text-to-speech via speakers to be turned on and off as needed.
* Beta: Added a [collectInputs](https://stripe.com/docs/terminal/features/collect-inputs) method to display forms and collect information from customers.

## Testing instructions

Used https://github.com/woocommerce/woocommerce-ios/pull/11080 for reference

Test any combination of card reader connection, and payment flows that worked previously using card present or tap to pay. The same behaviour should be expected after the update. As a guide, on my testing I went through the following scenarios:

* Connection from Order payment flow
* Connection from Payments menu
* Connection of a reader with a mandatory update
* Connection of M2 reader
* Connection from switched off state

I was not able to test the following, since I don't have the option:
* Connection of Chipper reader
* Take Interac payment
* Refund Interac payment
* Connection of WisePad 3 reader
* Take payment after in-line connection process
* Take payment with a previously-connected reader
* Take contactless payment
* Take card-inserted payment
* Reader update: error message when attempted with low battery
* Reader update: successfully updates software and starts payment (in that flow) when completed

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Podfile</code> was changed without updating its corresponding <code>Podfile.lock</code>. Please run <code>bundle exec pod install</code>.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by bozidarsevo** [comment]
@iamgabrielma CI is failing (non related test), so maybe you can check this before the CI passes :D

File: Podfile.lock
Comment (iamgabrielma) [inline Podfile.lock:1]: The only change in `Podfile.lock` for this PR should be `StripeTerminal (~> 3.3.1)`, isn't it?

File: Hardware/Hardware/CardReader/CardBrand.swift [hunk 1]
```diff
@@ -60,7 +66,7 @@ extension CardBrand {
             return "diners"
         case .interac:
             return "interac"
-        case .unknown:
+        case .unknown, .unionPay, .eftposAu:
```
Comment (iamgabrielma) [inline Hardware/Hardware/CardReader/CardBrand.swift]: I believe we don't support it, but we seem to have an icon for the `.unionPay` on `WCPayCardBrand.unionpay`:
```
case .unionpay:
  return iconName(suffix: "unionpay")
```

File: Hardware/Hardware/CardReader/CardBrand.swift [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import Codegen
 
 /// The various card brands for a card.
-public enum CardBrand: String, CaseIterable, Codable, GeneratedFakeable {
+@frozen public enum CardBrand: String, CaseIterable, Codable, GeneratedFakeable {
```
Comment (iamgabrielma) [inline Hardware/Hardware/CardReader/CardBrand.swift:4]: I'm curious about the usage of `@frozen` here (never used it tbh 😅). Are we just telling the compiler that we're not adding more cases anytime soon?

File: Hardware/Hardware/CardReader/CardPresentTransactionDetails.swift [hunk 3]
```diff
@@ -28,14 +28,24 @@ public struct CardPresentTransactionDetails: Codable, Equatable, GeneratedFakeab
     /// (Only applicable to EMV payments) The authorization data from the card issuer.
     public let emvAuthData: String?
 
+    /// If this payment is from a card wallet, this contains the details of the card wallet.
+    public let wallet: Wallet?
+
+    /// Identifies which network this charge was processed on.
+    /// Contains SCPCardBrand represented as a nullable NSNumber.
+    /// Only available after confirming the PaymentIntent.
+    public let networkDouble: Double?
```
Comment (iamgabrielma) [inline Hardware/Hardware/CardReader/CardPresentTransactionDetails.swift]: What do you think if we rename this `networkDouble` for clarity? Something like `cardBrandAssociatedIdentifier` or something similar that refers to the card brand associated value as an ID? We can also add a link to https://stripe.dev/stripe-terminal-ios/docs/Enums/SCPCardBrand.html in the comment for future reference.

File: Hardware/HardwareTests/Mocks/MockStripeCardPresentDetails.swift
Comment (iamgabrielma) [inline Hardware/HardwareTests/Mocks/MockStripeCardPresentDetails.swift:1]: Do we also need additional tests on `CardPresentDetailsTests` to confirm that `wallet` and `network` properties are mocked as expected?

File: Hardware/Hardware/CardReader/CardBrand.swift (hunk 2 above)
Comment (bozidarsevo) [reply Hardware/Hardware/CardReader/CardBrand.swift:4]: Yep, that is one of the things https://docs.swift.org/swift-book/documentation/the-swift-programming-language/attributes/#frozen

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/189">17.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by bozidarsevo** [comment]
@toupper @joshheald could you two also take a look at the PR? Tnx

**Comment by toupper** [comment]
Sorry @bozidarsevo, I don't have the capacity or the context at the moment, but I rely on @iamgabrielma and @joshheald to check this one :)

**Review by iamgabrielma — APPROVED** [review]
LGTM 🚢 We should update the `release-notes` informing about the SDK updated before merging. 

Along with your cases I also tested for:

- ✅ Take payment after in-line connection process
- ✅ Take payment with a previously-connected reader
- ✅ Take contactless payment
- ✅ Take card-inserted payment
- ✅ Reader update: error message when attempted with low battery
- ✅ Reader update: successfully updates software and starts payment (in that flow) when completed
- ✅ Connection of WisePad 3 reader

For reference, what remains untested:
- Connection of Chipper reader
- Take Interac payment
- Refund Interac payment

I'm not sure why but I don't seem to see warnings that come from the Hardware project unless I open the file specifically, in this case there's a warning to handle the `unknown` case in the `PaymentMethod` extension via `@unknown default`:

Screenshot 2024-03-12 at 12 26 23

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/191">17.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**